pkg/openstack/routes.go

```diff
@@ -288,82 +457,161 @@ func (r *Routes) CreateRoute(ctx context.Context, clusterName string, nameHint s
+	origRules, err := openstackutil.GetSecurityGroupRules(r.network, rules.ListOpts{SecGroupID: r.nodeSecurityGroupId, Direction: string(rules.DirIngress)})
+	if err != nil {
+		return err
+	}
+	sgRuleForPodCidrFound := false
+	sgRuleForNodeAddressFound := false
+	nodeAddrCidr := fmt.Sprintf("%s/32", addr)
+	for _, rule := range origRules {
+		if rule.RemoteIPPrefix == route.DestinationCIDR {
+			sgRuleForPodCidrFound = true
+		}
+		if rule.RemoteGroupID == r.nodeSecurityGroupId || rule.RemoteIPPrefix == nodeAddrCidr {
+			sgRuleForNodeAddressFound = true
+		}
+		if sgRuleForPodCidrFound && sgRuleForNodeAddressFound {
+			break
+		}
+	}
+	if !sgRuleForPodCidrFound {
+		// add security group rule for the pods of this node (route.TargetNode)
+		etherType := rules.EtherType4
+		if isCIDRv6 {
+			etherType = rules.EtherType6
+		}
+		unwind, err := createSecurityGroupRule(
+			r.network,
+			rules.CreateOpts{
+				SecGroupID:     r.nodeSecurityGroupId,
+				RemoteIPPrefix: route.DestinationCIDR,
+				Direction:      rules.DirIngress,
+				EtherType:      etherType})
 		if err != nil {
 			return err
 		}
-
 		defer onFailure.call(unwind)
-	} else {
-		// atomic route update
-		route := []routers.Route{{
-			DestinationCIDR: route.DestinationCIDR,
-			NextHop:         addr,
-		}}
-		unwind, err := addRoute(r.network, r.os.routeOpts.RouterID, route)
+	}
+	if !sgRuleForNodeAddressFound {
+		// add security group rule for this node (route.TargetNode)
+		etherType := rules.EtherType4
+		if isCIDRv6 {
+			etherType = rules.EtherType6
+		}
+		unwind, err := createSecurityGroupRule(
+			r.network,
+			rules.CreateOpts{
+				SecGroupID:     r.nodeSecurityGroupId,
+				RemoteIPPrefix: nodeAddrCidr,
+				Direction:      rules.DirIngress,
+				EtherType:      etherType})
 		if err != nil {
 			return err
 		}
-
 		defer onFailure.call(unwind)
 	}
 
-	if !r.allowedAddressPairs {
-		klog.V(4).Infof("Route created (skipping the allowed_address_pairs update): %v", route)
-		onFailure.disarm()
-		return nil
-	}
-
 	// get the port of addr on target node.
 	port, err := getPortByIP(r.network, addr, r.networkIDs)
 	if err != nil {
 		return err
 	}
+
 	if !port.PortSecurityEnabled {
-		klog.Warningf("Skipping allowed_address_pair for port: %s", port.ID)
+		klog.Warningf("Skipping updation of the port : %s (allowed_address_pairs and node security_group)", port.ID)
 		onFailure.disarm()
 		return nil
 	}
 
-	found := false
-	for _, item := range port.AllowedAddressPairs {
-		if item.IPAddress == route.DestinationCIDR {
-			klog.V(4).Infof("Found existing allowed-address-pair: %v", item)
-			found = true
-			break
+	if !r.allowedAddressPairs {
+		klog.V(4).Infof("Route created (skipping the allowed_address_pairs update): %v", route)
+	} else {
+		// update node port's AllowedAddressPairs, so that the traffic access the pods/services of this
+		// node (route.TargetNode) can enter into this node. in other words, permitting the packets the
```

Comment:
`AllowedAddressPairs` is about traffic leaving the node, not entering, I think.

Reply:
Yep, you are right.  I'll update it.